server/service/src/plugin/plugin_validation.md

```diff
@@ -0,0 +1,72 @@
+# Plugin Validation
+
+Plugins can modify the behaviour of mSupply and can potentially leak confidential information.
+For this reason its important that plugins come from a trusted party.
+For example, a malicious user could tricky an mSupply user to install a modified plugin.
+To avoid this, plugins are signed and only plugins that can be verified to come from a trusted party are loaded.
+
+To sign a plugin, a trusted party needs a private key for signing a plugin and a matching public certificate which is later used for verifying that the plugin hasn't been modified.
+The public certificate can either be self-signed or signed by another trusted party.
+Using this key pair, a plugin can be signed using the mSupply cli.
+A signed plugin contains a `manifest.json` and `manifest.signature` file.
+Besides others, the `manifest.json` file contains a list of plugin files and their hashes.
+Moreover, it contains the public certificate.
+
+To validate a plugin, the mSupply remote server needs to validate that it trusts the certificate that comes with the plugin (which is stored in the `manifest.json`).
+This is done by verifying the certificate against a list of trusted certificates which are stored in the `app_data/plugin_certs` directory.
+If the plugin certificate is trusted, it is then validated that the `manifest.json` file matches the `manifest.signature`.
+The `manifest.json` file is then used to only server plugin files that are listed in the manifest.
+
+# Generating Keys
+
+There are two types of public validation certificates.
+
+1. Self-singed certificates: These self-signed certificates need to be copied to the `app_data/plugin_certs`
+2. Certificates that are signed by a trusted party: These certificated must be signed by a certificate already in the `app_data/plugin_certs` server directory
+
+## Self-signed Certificates
+
+A self-signed certificate can, for example, be generated using:
+
+```bash
+openssl req -x509 -newkey rsa:2048 -keyout private.pem -out public.pem -nodes
+```
+
+In this example, the `public.pem` file needs to be copied to the `app_data/plugin_certs` server directory.
+
+## Certificates that are signed by a trusted party
+
+A 3rd party plugin developer can generate their own keypair and then let sign their certificate by mSupply:
```

Comment:
```suggestion
A 3rd party plugin developer can generate their own private key & signing request and then let mSupply generate a public key. The public key is then added to the list of valid decryption keys:
```

Reply:
Thanks, giving some more details now and changed it to:
```
A 3rd party plugin developer can generate their own private key & signing request and then let mSupply generate a public certificate.
This 3rd party certificate can be used to sign the 3rd party plugin but doesn't need to be added to the remote server.
However, the remote server needs to have the mSupply certificate installed (the one used to sign the 3rd party certificate).
I.e., the mSupply "root" certificate is used to validates the 3rd party certificate shipped with the plugin which in turn is used to validate that the plugin hasn't been modified.

```